pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -472,20 +472,93 @@ class StateVectorKokkos final
-        const ControlledGateOperation gateop =
-            reverse_lookup(controlled_gate_names, std::string_view{opName});
-        applyNCNamedOperation<KokkosExecSpace>(
-            gateop, *data_, num_qubits, controlled_wires, controlled_values,
-            wires, inverse, params);
+    /**
+     * @brief Apply a controlled-multi qubit operator to the state vector using
+     * a matrix
+     *
+     * @param matrix Kokkos gate matrix in the device space.
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (true or false).
+     * @param wires Wires to apply gate to.
+     * @param inverse Indicates whether to use adjoint of gate. (Default to
+     * false)
+     */
+    void applyNCMultiQubitOp(const KokkosVector matrix,
+                             const std::vector<std::size_t> &controlled_wires,
+                             const std::vector<bool> &controlled_values,
+                             const std::vector<std::size_t> &wires,
+                             bool inverse = false) {
+
+        auto &&num_qubits = this->getNumQubits();
+        std::size_t two2N =
+            std::exp2(num_qubits - wires.size() - controlled_wires.size());
+        std::size_t dim = std::exp2(wires.size());
+        KokkosVector matrix_trans("matrix_trans", matrix.size());
+
+        if (inverse) {
+            Kokkos::MDRangePolicy<DoubleLoopRank> policy_2d({0, 0}, {dim, dim});
+            Kokkos::parallel_for(
+                policy_2d,
+                KOKKOS_LAMBDA(const std::size_t i, const std::size_t j) {
+                    matrix_trans(i + j * dim) =
+                        Kokkos::conj(matrix(i * dim + j));
+                });
+        } else {
+            matrix_trans = matrix;
+        }
+
+        switch (wires.size()) {
+        case 1:
+            Kokkos::parallel_for(two2N, applyNC1QubitOpFunctor<fp_t>(
+                                            *data_, num_qubits, matrix_trans,
+                                            controlled_wires, controlled_values,
+                                            wires));
+            break;
+        case 2:
+            Kokkos::parallel_for(two2N, applyNC2QubitOpFunctor<fp_t>(
+                                            *data_, num_qubits, matrix_trans,
+                                            controlled_wires, controlled_values,
+                                            wires));
+            break;
+        case 3:
+            Kokkos::parallel_for(two2N, applyNC3QubitOpFunctor<fp_t>(
+                                            *data_, num_qubits, matrix_trans,
+                                            controlled_wires, controlled_values,
+                                            wires));
+            break;
+        default:
+            std::size_t scratch_size = ScratchViewComplex::shmem_size(dim) +
+                                       ScratchViewSizeT::shmem_size(dim);
```

Comment:
any concern on the shmem usage here?

Reply:
So this is actually already used in `applyMultiQubitOp` as well. I just tested on ISAIC A100 and here are the results:
- Level 0 (current implementation) fails for qubit unitary for >10 wires
- Level 1: does not fail (tested up to 14 wires, which takes a long time). There is a slight (at most few percent) time difference for <=10 wires. 
 
I will change the implementation to use level 1 scratchmemory instead of level 0, unless we disagree?

Reply:
Thanks @josephleekl for the testing. Sounds like there is some room to improve the usage of shmem to me. However, not a blocker for this PR. Could you please add an explore TODO ? We can come back to it later.

Reply:
No problem, I can add a TODO. There's definitely room to investigate the effect of this scratch memory on shmem for GPU and cache for CPU!